installinator-api/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! The REST API that installinator is a client of.
+//!
+//! Note that most of our APIs are named by their server. This one is instead
+//! named by the client, since it is expected that multiple services will
+//! implement it.
+
+use anyhow::{anyhow, Result};
+use dropshot::{
+    ConfigDropshot, FreeformBody, HandlerTaskMode, HttpError,
+    HttpResponseHeaders, HttpResponseOk, HttpResponseUpdatedNoContent,
+    HttpServerStarter, Path, RequestContext, TypedBody,
+};
+use hyper::{header, Body, StatusCode};
+use installinator_common::EventReport;
+use omicron_common::update::ArtifactHashId;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use uuid::Uuid;
+
+#[derive(Debug, Deserialize, JsonSchema)]
+pub struct ReportQuery {
+    /// A unique identifier for the update.
+    pub update_id: Uuid,
+}
+
+#[dropshot::api_description]
+pub trait InstallinatorApi {
+    type Context;
+
+    /// Fetch an artifact by hash.
+    #[endpoint {
+        method = GET,
+        path = "/artifacts/by-hash/{kind}/{hash}",
+    }]
+    async fn get_artifact_by_hash(
+        rqctx: RequestContext<Self::Context>,
+        path: Path<ArtifactHashId>,
+    ) -> Result<HttpResponseHeaders<HttpResponseOk<FreeformBody>>, HttpError>;
+
+    /// Report progress and completion to the server.
+    ///
+    /// This method requires an `update_id` path parameter. This update ID is
+    /// matched against the server currently performing an update. If the
+    /// server is unaware of the update ID, it will return an HTTP 422
+    /// Unprocessable Entity code.
+    #[endpoint {
+        method = POST,
+        path = "/report-progress/{update_id}",
+    }]
+    async fn report_progress(
+        rqctx: RequestContext<Self::Context>,
+        path: Path<ReportQuery>,
+        report: TypedBody<EventReport>,
+    ) -> Result<HttpResponseUpdatedNoContent, HttpError>;
+}
+
+/// Add a content length header to a response.
+///
+/// Intended to be called by `get_artifact_by_hash` implementations.
+pub fn body_to_artifact_response(
+    size: u64,
+    body: Body,
+) -> HttpResponseHeaders<HttpResponseOk<FreeformBody>> {
+    let mut response =
+        HttpResponseHeaders::new_unnamed(HttpResponseOk(body.into()));
+    let headers = response.headers_mut();
+    headers.append(header::CONTENT_LENGTH, size.into());
+    response
+}
+
+/// The result of processing an installinator event report.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Hash)]
+#[must_use]
+pub enum EventReportStatus {
+    /// This report was processed by the server.
+    Processed,
+
+    /// The update ID was not recognized by the server.
+    UnrecognizedUpdateId,
+
+    /// The progress receiver is closed.
+    ReceiverClosed,
+}
+
+impl EventReportStatus {
+    /// Convert this status to an HTTP result.
+    ///
+    /// Intended to be called by `report_progress` implementations.
+    pub fn to_http_result(
```

Comment:
I could have turned this logic into living on the trait, but in this case that extra work seems unnecessary.